File: challenger-backend/src/main/java/com/kameo/challenger/odb/TaskODB.java
package com.kameo.challenger.odb;

import com.kameo.challenger.odb.api.IIdentity;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import javax.persistence.*;
import javax.validation.constraints.NotNull;
import java.util.Date;


@NoArgsConstructor
@ToString(of = "id")
@Entity
@Table(indexes = {
        @Index(columnList = "user_id"),
        @Index(columnList = "challenge_id"),
})
@Data
public class TaskODB implements IIdentity {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private long id;
    @Enumerated
    @NotNull
    private TaskType taskType;
    private String icon;
    private int difficulty;
    private String label;
    @Enumerated
    @NotNull
    private TaskStatus taskStatus;
    @Temporal(TemporalType.TIMESTAMP)
    private Date dueDate;
    @ManyToOne
    @NotNull
    private UserODB user;
    @ManyToOne
    @NotNull
    private UserODB createdByUser;

    @Transient
    boolean done;

    @NotNull
    @ManyToOne
    private ChallengeODB challenge;

    public TaskODB(long id) {
        this.id = id;
    }

    public ChallengeODB getChallenge() {
        return challenge;
    }

    public void setChallenge(ChallengeODB challenge) {
        this.challenge = challenge;
    }

    public UserODB getUser() {
        return user;
    }

    public void setUser(UserODB user) {
        this.user = user;
    }

    public TaskStatus getTaskStatus() {
        return taskStatus;
    }

    public TaskType getTaskType() {
        return taskType;
    }

    public void setTaskStatus(TaskStatus taskStatus) {
        this.taskStatus = taskStatus;
    }

    public void setTaskType(TaskType taskType) {
        this.taskType = taskType;
    }

    public boolean isDone() {
        return done;
    }

    public void setDone(boolean done) {
        this.done = done;
    }

    public Date getDueDate() {
        return dueDate;
    }

    public void setDueDate(Date dueDate) {
        this.dueDate = dueDate;
    }

    @Override
    public long getId() {
        return id;
    }

    @Override
    public boolean isNew() {
        return getId()<0;
    }
}

File: challenger-backend/src/main/java/com/kameo/challenger/domain/tasks/db/TaskODB.kt
package com.kameo.challenger.domain.tasks.db

import com.kameo.challenger.domain.accounts.db.UserODB
import com.kameo.challenger.domain.challenges.db.ChallengeODB
import com.kameo.challenger.odb.api.IIdentity
import java.time.LocalDate
import java.time.LocalDateTime
import javax.persistence.*


@Entity
data class TaskODB(@Id
                   @GeneratedValue(strategy = javax.persistence.GenerationType.AUTO)
                   override var id: Long = 0) : IIdentity {


// var num: Int=10

    @Enumerated
    lateinit var taskType: TaskType

    var difficulty: Int = 0

    lateinit var label: String

    @Enumerated
    lateinit var taskStatus: TaskStatus


    @ManyToOne
    lateinit var user: UserODB

    @ManyToOne
    lateinit var createdByUser: UserODB


    var createDate: LocalDateTime = LocalDateTime.now()

    @ManyToOne
    lateinit var challenge: ChallengeODB


    var icon: String? = null


    // from when task will be visible
    var startDate: LocalDateTime? = null
    var dueDate: LocalDateTime? = null



    var closeDate: LocalDate? = null

    @Transient
    var done: Boolean = false

    var monthDays: String? = null
    var weekDays: String? = null



    enum class TaskDifficulty {
        EASY,
        MEDIUM,
        HARD
    }

}
